source/common/router/scoped_rds.cc

```diff
@@ -143,13 +143,17 @@ void ScopedRdsConfigSubscription::onConfigUpdate(
   for (const auto& scope_name : removed_resources) {
     auto iter = scoped_route_map_.find(scope_name);
     if (iter != scoped_route_map_.end()) {
+      ScopedRouteInfoConstSharedPtr to_be_deleted = iter->second;
       scope_name_by_hash_.erase(iter->second->scopeKey().hash());
       scoped_route_map_.erase(iter);
-      applyDeltaConfigUpdate([scope_name](const ConfigProvider::ConfigConstSharedPtr& config) {
-        auto* thread_local_scoped_config = const_cast<ThreadLocalScopedConfigImpl*>(
-            static_cast<const ThreadLocalScopedConfigImpl*>(config.get()));
-        thread_local_scoped_config->removeRoutingScope(scope_name);
-      });
+      applyDeltaConfigUpdate(
+          [scope_name](const ConfigProvider::ConfigConstSharedPtr& config) {
+            auto* thread_local_scoped_config = const_cast<ThreadLocalScopedConfigImpl*>(
+                static_cast<const ThreadLocalScopedConfigImpl*>(config.get()));
+            thread_local_scoped_config->removeRoutingScope(scope_name);
+          },
+          // We need to delete the associated RouteConfigProvider in main thread.
+          [to_be_deleted]() { /*to_be_deleted is destructed in main thread.*/ });
```

Comment:
I suggest adding more documentation about why the deletion needs to be deferred to the main thread. This may not be as obvious to someone who is not familiar with the ConfigProvider framework and/or the `applyDeltaConfigUpdate()` implementation.

Reply:
done. 
It's actually related to the fact that RouteConfigProvider contains a thread local slot which can only be destructed in main thread, which adds a somewhat weird constraint that a RouteConfigProvider instance can only be deleted in main thread.